

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API reference &mdash; ScoriaPy 0.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=9edc463e" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=7026087e"></script>
      <script src="_static/doctools.js?v=fd6eb6e6"></script>
      <script src="_static/sphinx_highlight.js?v=6ffebe34"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="ScoriaPy documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            ScoriaPy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">API reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-scoriapy.pp">Preprocessing (<code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.pp</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pp.preprocess_raw_and_normalize"><code class="docutils literal notranslate"><span class="pre">preprocess_raw_and_normalize()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pp.read_and_concat"><code class="docutils literal notranslate"><span class="pre">read_and_concat()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pp.run_hvg"><code class="docutils literal notranslate"><span class="pre">run_hvg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pp.run_scanpy_basic_pipeline"><code class="docutils literal notranslate"><span class="pre">run_scanpy_basic_pipeline()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scoriapy.tl">Analysis tools (<code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.tl</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.aggregate_group_expression"><code class="docutils literal notranslate"><span class="pre">aggregate_group_expression()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.aggregate_single_cell_data"><code class="docutils literal notranslate"><span class="pre">aggregate_single_cell_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.check_group_columns_exist"><code class="docutils literal notranslate"><span class="pre">check_group_columns_exist()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.combine_group_values"><code class="docutils literal notranslate"><span class="pre">combine_group_values()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.compare_cluster_phase"><code class="docutils literal notranslate"><span class="pre">compare_cluster_phase()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.compare_gene_expression"><code class="docutils literal notranslate"><span class="pre">compare_gene_expression()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.compare_gene_expression_two"><code class="docutils literal notranslate"><span class="pre">compare_gene_expression_two()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.create_group_dicts"><code class="docutils literal notranslate"><span class="pre">create_group_dicts()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.detect_gene_flips"><code class="docutils literal notranslate"><span class="pre">detect_gene_flips()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.get_de_genes"><code class="docutils literal notranslate"><span class="pre">get_de_genes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.get_deg_df"><code class="docutils literal notranslate"><span class="pre">get_deg_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.get_pairwise_deg"><code class="docutils literal notranslate"><span class="pre">get_pairwise_deg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.make_pseudobulk_adata"><code class="docutils literal notranslate"><span class="pre">make_pseudobulk_adata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.run_anova_with_posthoc"><code class="docutils literal notranslate"><span class="pre">run_anova_with_posthoc()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.run_jasmine"><code class="docutils literal notranslate"><span class="pre">run_jasmine()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.run_rank_genes_groups"><code class="docutils literal notranslate"><span class="pre">run_rank_genes_groups()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.run_resolution_sweep_parallel"><code class="docutils literal notranslate"><span class="pre">run_resolution_sweep_parallel()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.subset_and_clean_adata"><code class="docutils literal notranslate"><span class="pre">subset_and_clean_adata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.tl.summarize_category_distribution"><code class="docutils literal notranslate"><span class="pre">summarize_category_distribution()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scoriapy.pl">Plotting (<code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.pl</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.add_ncells_annotation"><code class="docutils literal notranslate"><span class="pre">add_ncells_annotation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.add_text_labels"><code class="docutils literal notranslate"><span class="pre">add_text_labels()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.calculate_median_position"><code class="docutils literal notranslate"><span class="pre">calculate_median_position()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.create_heatmap_annotation"><code class="docutils literal notranslate"><span class="pre">create_heatmap_annotation()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.get_heatmap_dfs"><code class="docutils literal notranslate"><span class="pre">get_heatmap_dfs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_cluster_heatmap"><code class="docutils literal notranslate"><span class="pre">plot_cluster_heatmap()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_distribution"><code class="docutils literal notranslate"><span class="pre">plot_distribution()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_embeddings_with_labels"><code class="docutils literal notranslate"><span class="pre">plot_embeddings_with_labels()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_group_embeddings"><code class="docutils literal notranslate"><span class="pre">plot_group_embeddings()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_pca_for_samples"><code class="docutils literal notranslate"><span class="pre">plot_pca_for_samples()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_umap_by_categories"><code class="docutils literal notranslate"><span class="pre">plot_umap_by_categories()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_umap_by_categories_v1"><code class="docutils literal notranslate"><span class="pre">plot_umap_by_categories_v1()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_umap_by_group"><code class="docutils literal notranslate"><span class="pre">plot_umap_by_group()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_umap_grid_by_sample"><code class="docutils literal notranslate"><span class="pre">plot_umap_grid_by_sample()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_umap_param_grid"><code class="docutils literal notranslate"><span class="pre">plot_umap_param_grid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.plot_violin_genes"><code class="docutils literal notranslate"><span class="pre">plot_violin_genes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.preserve_leiden_categories"><code class="docutils literal notranslate"><span class="pre">preserve_leiden_categories()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.sort_cells_by_metadata"><code class="docutils literal notranslate"><span class="pre">sort_cells_by_metadata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.pl.subset_adata"><code class="docutils literal notranslate"><span class="pre">subset_adata()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scoriapy.utils">Utilities (<code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.utils</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.FigureRegistry"><code class="docutils literal notranslate"><span class="pre">FigureRegistry</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#scoriapy.utils.FigureRegistry.add"><code class="docutils literal notranslate"><span class="pre">FigureRegistry.add()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#scoriapy.utils.FigureRegistry.save_all"><code class="docutils literal notranslate"><span class="pre">FigureRegistry.save_all()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#scoriapy.utils.FigureRegistry.show"><code class="docutils literal notranslate"><span class="pre">FigureRegistry.show()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.confirm_and_set_colors"><code class="docutils literal notranslate"><span class="pre">confirm_and_set_colors()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.create_cell_mask"><code class="docutils literal notranslate"><span class="pre">create_cell_mask()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.display_colors"><code class="docutils literal notranslate"><span class="pre">display_colors()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.display_colors_with_keys"><code class="docutils literal notranslate"><span class="pre">display_colors_with_keys()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.extract_cell_cluster"><code class="docutils literal notranslate"><span class="pre">extract_cell_cluster()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.get_full_df"><code class="docutils literal notranslate"><span class="pre">get_full_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.get_mat_df"><code class="docutils literal notranslate"><span class="pre">get_mat_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.make_annotation_df"><code class="docutils literal notranslate"><span class="pre">make_annotation_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.make_mat_dfs"><code class="docutils literal notranslate"><span class="pre">make_mat_dfs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.map_colors"><code class="docutils literal notranslate"><span class="pre">map_colors()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.scale_df"><code class="docutils literal notranslate"><span class="pre">scale_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.setup_logger"><code class="docutils literal notranslate"><span class="pre">setup_logger()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.subset_adata_by_mask"><code class="docutils literal notranslate"><span class="pre">subset_adata_by_mask()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scoriapy.utils.subset_by_markers"><code class="docutils literal notranslate"><span class="pre">subset_by_markers()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ScoriaPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API reference</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api-reference">
<h1>API reference<a class="headerlink" href="#api-reference" title="Link to this heading"></a></h1>
<p>This section documents the public API of the <a class="reference internal" href="generated/scoriapy.html#module-scoriapy" title="scoriapy"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy</span></code></a> package.
The modules below are organized following the Scanpy convention.</p>
<section id="module-scoriapy.pp">
<span id="preprocessing-scoriapy-pp"></span><h2>Preprocessing (<a class="reference internal" href="generated/scoriapy.pp.html#module-scoriapy.pp" title="scoriapy.pp"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.pp</span></code></a>)<a class="headerlink" href="#module-scoriapy.pp" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pp.preprocess_raw_and_normalize">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pp.</span></span><span class="sig-name descname"><span class="pre">preprocess_raw_and_normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pp.preprocess_raw_and_normalize" title="Link to this definition"></a></dt>
<dd><p>Preprocess an AnnData object by optionally restoring <cite>.raw</cite>,
normalizing total counts, and applying log1p transformation.</p>
<p>This function is safe and self-contained: it always works on a copy
of the input AnnData object and therefore does not modify the original.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The input AnnData object to preprocess.</p></li>
<li><p><strong>use_raw</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – If True and <cite>adata.raw</cite> exists, the <cite>.raw</cite> stored data
is restored via <cite>adata.raw.to_adata()</cite> before normalization.
If False, the existing <cite>.X</cite> matrix is used as-is.</p></li>
<li><p><strong>target_sum</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1e4</em><em>)</em>) – Target sum of total counts per cell for <cite>sc.pp.normalize_total</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A processed AnnData object with:
- raw restored (if available and <cite>use_raw=True</cite>)
- normalized counts (<cite>sc.pp.normalize_total</cite>)
- log1p-transformed values (<cite>sc.pp.log1p</cite>)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pp.read_and_concat">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pp.</span></span><span class="sig-name descname"><span class="pre">read_and_concat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">process_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logger</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">concat_kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pp.read_and_concat" title="Link to this definition"></a></dt>
<dd><p>Load multiple samples using a custom processing function and concatenate them.</p>
<p>This orchestrator applies a user-defined loading function to each sample path
in a dictionary and merges the results into a single AnnData object using
the modern <code class="docutils literal notranslate"><span class="pre">anndata.concat</span></code> API.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em> | </em><em>Path</em><em>]</em>) – Mapping of sample names to their respective file paths.</p></li>
<li><p><strong>process_func</strong> (<em>Callable</em><em>[</em><em>[</em><em>Path</em><em>, </em><em>str</em><em>]</em><em>, </em><em>AnnData</em><em>]</em>) – A callable that takes (Path, str) and returns an AnnData object.
Highly recommended to use <code class="docutils literal notranslate"><span class="pre">functools.partial</span></code> to pass extra
metadata or project settings.</p></li>
<li><p><strong>logger</strong> (<em>Logger</em><em> | </em><em>None</em>) – A <code class="xref py py-class docutils literal notranslate"><span class="pre">logging.Logger</span></code> object for progress tracking.
If None, a basic console logger is initialized.</p></li>
<li><p><strong>**concat_kwargs</strong> – Keyword arguments passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">anndata.concat()</span></code>.
Commonly used: <code class="docutils literal notranslate"><span class="pre">join='inner'</span></code>, <code class="docutils literal notranslate"><span class="pre">uns_merge='unique'</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Concatenated AnnData object with <code class="docutils literal notranslate"><span class="pre">.obs['sample_name']</span></code> populated.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ad.AnnData</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">partial</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># 1. Define metadata and sample paths</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">meta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;age&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;6m&quot;</span><span class="p">,</span> <span class="s2">&quot;24m&quot;</span><span class="p">]},</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="s2">&quot;S2&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">paths</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;S1&quot;</span><span class="p">:</span> <span class="s2">&quot;data/s1.h5&quot;</span><span class="p">,</span> <span class="s2">&quot;S2&quot;</span><span class="p">:</span> <span class="s2">&quot;data/s2.h5&quot;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># 2. Define a loader that requires the metadata table</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span><span class="w"> </span><span class="nf">my_loader</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">info_df</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">info_df</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">info_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
<span class="gp">... </span>            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">name</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">adata</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># 3. Create a partial function to &#39;bake in&#39; the metadata</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loader_with_meta</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">my_loader</span><span class="p">,</span> <span class="n">info_df</span><span class="o">=</span><span class="n">meta</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># 4. Run orchestrator</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">read_and_concat</span><span class="p">(</span><span class="n">paths</span><span class="p">,</span> <span class="n">loader_with_meta</span><span class="p">,</span> <span class="n">join</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pp.run_hvg">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pp.</span></span><span class="sig-name descname"><span class="pre">run_hvg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flavor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seurat'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pp.run_hvg" title="Link to this definition"></a></dt>
<dd><p>Compute and subset highly variable genes (HVGs).</p>
<p>This is a thin wrapper around <code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.pp.highly_variable_genes()</span></code> that
logs progress and returns a view restricted to the selected genes. If the
data has not been log-transformed yet (i.e. <code class="docutils literal notranslate"><span class="pre">&quot;log1p&quot;</span></code> is not present in
<code class="docutils literal notranslate"><span class="pre">adata.uns</span></code>), a <code class="docutils literal notranslate"><span class="pre">log1p</span></code> transform is applied before selecting HVGs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Input AnnData object.</p></li>
<li><p><strong>n_top</strong> (<em>int</em>) – Number of highly variable genes to keep.</p></li>
<li><p><strong>flavor</strong> (<em>str</em>) – HVG selection flavor passed to Scanpy (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;seurat&quot;</span></code>).</p></li>
<li><p><strong>log</strong> – Optional <code class="xref py py-class docutils literal notranslate"><span class="pre">logging.Logger</span></code> instance. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, a default
scoriapy logger is created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Copy of <code class="docutils literal notranslate"><span class="pre">adata</span></code> restricted to the selected HVGs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pp.run_scanpy_basic_pipeline">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pp.</span></span><span class="sig-name descname"><span class="pre">run_scanpy_basic_pipeline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logger</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pp.run_scanpy_basic_pipeline" title="Link to this definition"></a></dt>
<dd><p>Run standard Scanpy preprocessing with logging at each step:
HVG → scale → PCA → neighbors → UMAP → Leiden.</p>
<p>This pipeline automates the transition from raw/normalized counts to a
clustered embedding. It handles high-variable gene selection, data scaling
to unit variance, dimensionality reduction via PCA, and manifold learning
(UMAP/Leiden).</p>
<dl>
<dt>Args:</dt><dd><p>adata (AnnData): Annotated data matrix (typically normalized and log-transformed).
n_top_genes (int): Number of highly variable genes to keep. Default is 2000.
n_pcs (int): Number of principal components to compute and use for neighbors.
n_neighbors (int): Size of the local neighborhood for the KNN graph.
resolution (float): Resolution parameter for Leiden clustering (higher = more clusters).
hvg (bool): If True, subsets the object to highly variable genes.
scale (bool): If True, scales data to unit variance and zero mean.
copy (bool): If True, returns a new AnnData object; otherwise modifies in-place.
logger (logging.Logger, optional): Logger object for progress tracking.</p>
<blockquote>
<div><p>If None, calls setup_logger().</p>
</div></blockquote>
</dd>
<dt>Returns:</dt><dd><p>AnnData: The processed AnnData object with PCA, UMAP, and Leiden results.</p>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load raw data and perform basic normalization</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Run the automated pipeline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span> <span class="o">=</span> <span class="n">setup_logger</span><span class="p">(</span><span class="s2">&quot;preprocess.log&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata_processed</span> <span class="o">=</span> <span class="n">run_scanpy_basic_pipeline</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">logger</span><span class="o">=</span><span class="n">logger</span>
<span class="gp">... </span>    <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">resolution</span><span class="o">=</span><span class="mf">0.8</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Visualize the results</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_processed</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scoriapy.tl">
<span id="analysis-tools-scoriapy-tl"></span><h2>Analysis tools (<a class="reference internal" href="generated/scoriapy.tl.html#module-scoriapy.tl" title="scoriapy.tl"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.tl</span></code></a>)<a class="headerlink" href="#module-scoriapy.tl" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.aggregate_group_expression">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">aggregate_group_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_mask</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.aggregate_group_expression" title="Link to this definition"></a></dt>
<dd><p>Aggregate gene expression for a specific group.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with expression data.</p></li>
<li><p><strong>group_mask</strong> – Index or boolean mask selecting cells belonging to the group.</p></li>
<li><p><strong>group_name</strong> (<em>str</em>) – Name to use for the aggregated pseudo-bulk sample.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Single-column DataFrame indexed by gene names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.aggregate_single_cell_data">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">aggregate_single_cell_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.aggregate_single_cell_data" title="Link to this definition"></a></dt>
<dd><p>Aggregate single-cell expression data by the given group columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with expression data.</p></li>
<li><p><strong>group_cols</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of <code class="docutils literal notranslate"><span class="pre">obs</span></code> columns to group by (e.g. <code class="docutils literal notranslate"><span class="pre">[&quot;sample&quot;,</span> <span class="pre">&quot;cell_type&quot;]</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pseudo-bulk expression matrix with groups as columns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.check_group_columns_exist">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">check_group_columns_exist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.check_group_columns_exist" title="Link to this definition"></a></dt>
<dd><p>Ensure that all requested grouping columns exist in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object whose <code class="docutils literal notranslate"><span class="pre">obs</span></code> will be checked.</p></li>
<li><p><strong>group_cols</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of column names that must be present in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If any of the requested columns are missing.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.combine_group_values">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">combine_group_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">group_values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.combine_group_values" title="Link to this definition"></a></dt>
<dd><p>Combine group column values into a single string, separated by underscores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>group_values</strong> – Iterable of values describing a group.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Combined group name.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.compare_cluster_phase">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">compare_cluster_phase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cross_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'age'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.compare_cluster_phase" title="Link to this definition"></a></dt>
<dd><p>Apply a Kruskal-Wallis test to each phase for a given cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cross_table</strong> (<em>DataFrame</em>) – Cross-tabulated table (e.g. from <code class="xref py py-func docutils literal notranslate"><span class="pre">pandas.crosstab()</span></code>) whose index
includes the cluster identifier.</p></li>
<li><p><strong>cluster_id</strong> (<em>str</em>) – Specific cluster label to analyze.</p></li>
<li><p><strong>level</strong> (<em>int</em>) – Level of the index in <code class="docutils literal notranslate"><span class="pre">cross_table</span></code> corresponding to <code class="docutils literal notranslate"><span class="pre">cluster_id</span></code>.</p></li>
<li><p><strong>group_level</strong> (<em>str</em>) – Index level used to define groups for the test (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;age&quot;</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapping from phase (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;G1&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;G2M&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;S&quot;</span></code>) to
Kruskal-Wallis test results.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.compare_gene_expression">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">compare_gene_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby_col</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.compare_gene_expression" title="Link to this definition"></a></dt>
<dd><p>Compare expression of a single gene across multiple groups with Kruskal-Wallis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with expression data.</p></li>
<li><p><strong>clustering_key</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used to subset to specific clusters.</p></li>
<li><p><strong>cluster_ids</strong> – One or more cluster identifiers to include.</p></li>
<li><p><strong>gene</strong> (<em>str</em>) – Gene name to test.</p></li>
<li><p><strong>groupby_col</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> defining the groups to compare.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dunn’s test results as a DataFrame if the global test is significant,
otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.compare_gene_expression_two">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">compare_gene_expression_two</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby_col</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.compare_gene_expression_two" title="Link to this definition"></a></dt>
<dd><p>Compare expression of a single gene between exactly two groups using Mann-Whitney U.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with expression data.</p></li>
<li><p><strong>clustering_key</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used to subset to specific clusters.</p></li>
<li><p><strong>cluster_ids</strong> – One or more cluster identifiers to include.</p></li>
<li><p><strong>gene</strong> (<em>str</em>) – Gene name to test.</p></li>
<li><p><strong>groupby_col</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> defining the two groups being compared.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with keys <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;statistic&quot;</span></code>, and <code class="docutils literal notranslate"><span class="pre">&quot;p_value&quot;</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.create_group_dicts">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">create_group_dicts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.create_group_dicts" title="Link to this definition"></a></dt>
<dd><p>Create a list of dictionaries defining groups based on specified columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object whose <code class="docutils literal notranslate"><span class="pre">obs</span></code> is used to define groups.</p></li>
<li><p><strong>group_cols</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of column names to group by.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of dictionaries describing each group combination.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.detect_gene_flips">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">detect_gene_flips</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">one_down</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">one_up</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">two_down</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">two_up</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.detect_gene_flips" title="Link to this definition"></a></dt>
<dd><p>Identify genes that flip direction between two conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>one_down</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Gene lists for the two conditions.</p></li>
<li><p><strong>one_up</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Gene lists for the two conditions.</p></li>
<li><p><strong>two_down</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Gene lists for the two conditions.</p></li>
<li><p><strong>two_up</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Gene lists for the two conditions.</p></li>
<li><p><strong>return_df</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, return a DataFrame, otherwise a dict.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.get_de_genes">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">get_de_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'age'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.get_de_genes" title="Link to this definition"></a></dt>
<dd><p>Extract sets of differentially expressed genes for each group.</p>
<p>Assumes that differential expression results are already present in
<code class="docutils literal notranslate"><span class="pre">adata.uns[key]</span></code> as created by <code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.tl.rank_genes_groups()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with DE results.</p></li>
<li><p><strong>group_col</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that defines the groups.</p></li>
<li><p><strong>key</strong> (<em>str</em>) – Key under which DE results are stored in <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code>.</p></li>
<li><p><strong>query</strong> (<em>str</em><em> | </em><em>None</em>) – Optional pandas query string to filter the DE table for each group.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapping from group label to a set of gene names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.get_deg_df">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">get_deg_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby_attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.get_deg_df" title="Link to this definition"></a></dt>
<dd><p>Compute a DEG table for a given query and return expression for selected genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with expression data.</p></li>
<li><p><strong>cluster_key</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> indicating clusters (currently not used in
the implementation but kept for API compatibility).</p></li>
<li><p><strong>groupby_attribute</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used for grouping in DE analysis.</p></li>
<li><p><strong>rank_key</strong> (<em>str</em>) – Key under which to store DE results in <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code>.</p></li>
<li><p><strong>query</strong> (<em>str</em>) – Pandas query string used to filter the DE table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>test_df</em> – Expression matrix (genes x cells) for the selected genes.</p></li>
<li><p><em>filter_df</em> – DataFrame with statistics for the selected genes.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>Tuple</em>[<em>DataFrame</em>, <em>DataFrame</em>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.get_pairwise_deg">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">get_pairwise_deg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comparisons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'wilcoxon'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.get_pairwise_deg" title="Link to this definition"></a></dt>
<dd><p>Perform pairwise differential expression analysis for specified comparisons.</p>
<p>This is a convenience wrapper around <code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.tl.rank_genes_groups()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with expression data.</p></li>
<li><p><strong>comparisons</strong> (<em>Iterable</em><em>[</em><em>Tuple</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em><em>]</em>) – Iterable of pairs <code class="docutils literal notranslate"><span class="pre">(group1,</span> <span class="pre">group2)</span></code> to compare.</p></li>
<li><p><strong>groupby_col</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used for grouping.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Differential expression method passed to Scanpy (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;wilcoxon&quot;</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapping from comparison label (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;A_vs_B&quot;</span></code>) to a
<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> of DE statistics.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.make_pseudobulk_adata">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">make_pseudobulk_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cluster_name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('age',</span> <span class="pre">'sample_id')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sum'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">make_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.make_pseudobulk_adata" title="Link to this definition"></a></dt>
<dd><p>Create pseudobulk counts for one or more cell types using Scanpy aggregation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Single-cell AnnData.</p></li>
<li><p><strong>cell_type</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em>) – Cell type(s) to subset (e.g. “Myo”, “Fibro”, or [“AV”, “HS”] for Luminal).</p></li>
<li><p><strong>cell_type_col</strong> (<em>str</em>) – Column in adata.obs containing cell-type labels.</p></li>
<li><p><strong>groupby</strong> (<em>tuple</em>) – Variables to aggregate by (default: age, sample_id).</p></li>
<li><p><strong>layer_name</strong> (<em>str</em>) – Name of layer created by sc.get.aggregate (default “sum”).</p></li>
<li><p><strong>make_int</strong> (<em>bool</em>) – Convert counts to int32 for DESeq2.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pseudobulk AnnData with samples as rows and genes as columns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.run_anova_with_posthoc">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">run_anova_with_posthoc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iv</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.run_anova_with_posthoc" title="Link to this definition"></a></dt>
<dd><p>Performs a standard statistical pipeline: Levene’s Test for homogeneity
of variance, Type II ANOVA, and Tukey’s HSD post-hoc test.</p>
<dl>
<dt>Args:</dt><dd><p>df (pd.DataFrame): The dataset containing the data.
dv (str): The column name of the dependent variable (continuous).
iv (str): The column name of the independent variable (categorical).</p>
</dd>
<dt>Returns:</dt><dd><p>pd.DataFrame: A dataframe containing Tukey HSD pairwise comparisons
augmented with ANOVA and Levene test p-values.</p>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>    <span class="s1">&#39;diet&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;weight_loss&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">analyze_variance_pipeline</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">dv</span><span class="o">=</span><span class="s1">&#39;weight_loss&#39;</span><span class="p">,</span> <span class="n">iv</span><span class="o">=</span><span class="s1">&#39;diet&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.run_jasmine">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">run_jasmine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">net</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'oddsratio'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'jasmine'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.run_jasmine" title="Link to this definition"></a></dt>
<dd><p>JASMINE multi-signature enrichment scoring for single-cell RNA-seq.</p>
<p>This function computes a robust per-cell signature score using the
JASMINE method, which combines:</p>
<ol class="arabic simple">
<li><p>Rank component (dropout-robust)
Mean normalized rank of signature genes among expressed genes only.</p></li>
<li><p>Enrichment component
Odds ratio or likelihood enrichment of expressed signature genes
relative to background genes.</p></li>
</ol>
<p>Final score:</p>
<blockquote>
<div><p>score = (rank_component + enrichment_component) / 2</p>
</div></blockquote>
<p>Results are stored similarly to AUCell/Scanpy scoring.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Expression matrix of shape (n_cells × n_genes).</p></li>
<li><p><strong>net</strong> (<em>pandas.DataFrame</em>) – <p>Signature definition table with columns:</p>
<blockquote>
<div><p>source : signature name
target : gene symbol</p>
</div></blockquote>
<p>Example:</p>
<blockquote>
<div><p>source   target
Tcell    CD3D
Tcell    CD3E
Bcell    MS4A1</p>
</div></blockquote>
</p></li>
<li><p><strong>layer</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>default=None</em>) – Use adata.layers[layer] instead of adata.X.</p></li>
<li><p><strong>method</strong> (<em>{'oddsratio'</em><em>, </em><em>'likelihood'}</em>) – Enrichment calculation.</p></li>
<li><p><strong>key_added</strong> (<em>str</em>) – Store results in adata.obsm[key_added].</p></li>
<li><p><strong>copy</strong> (<em>bool</em>) – Return a modified copy (True) or update in place (False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>AnnData or None</em></p></li>
<li><p><em>Output</em></p></li>
<li><p><em>——</em></p></li>
<li><p><strong>adata.obsm[key_added]</strong> (<em>pandas.DataFrame</em>) – Shape = (cells × signatures)</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<dl class="simple">
<dt>Performance:</dt><dd><ul class="simple">
<li><p>Sparse CSR only</p></li>
<li><p>Numba-accelerated ranking</p></li>
<li><p>Matrix-multiply enrichment</p></li>
<li><p>20–100× faster than naive Python loops</p></li>
</ul>
</dd>
<dt>Complexity:</dt><dd><p>O(nnz + cells × signatures)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">jasmine</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_jasmine</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">net</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>    <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Tcell&quot;</span><span class="p">,</span><span class="s2">&quot;Tcell&quot;</span><span class="p">,</span><span class="s2">&quot;Bcell&quot;</span><span class="p">,</span><span class="s2">&quot;Bcell&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s2">&quot;target&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CD3D&quot;</span><span class="p">,</span><span class="s2">&quot;CD3E&quot;</span><span class="p">,</span><span class="s2">&quot;MS4A1&quot;</span><span class="p">,</span><span class="s2">&quot;CD79A&quot;</span><span class="p">]</span>
<span class="gp">... </span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">run_jasmine</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">net</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;jasmine&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;jasmine_Tcell&quot;</span><span class="p">,</span> <span class="s2">&quot;jasmine_Bcell&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.run_rank_genes_groups">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">run_rank_genes_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby_attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.run_rank_genes_groups" title="Link to this definition"></a></dt>
<dd><p>Run <code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.tl.rank_genes_groups()</span></code> and plot the result.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with expression data.</p></li>
<li><p><strong>groupby_attribute</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used to define groups.</p></li>
<li><p><strong>rank_key</strong> (<em>str</em><em> | </em><em>None</em>) – Optional key under which to store results in <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code>. If
<code class="docutils literal notranslate"><span class="pre">None</span></code>, defaults to <code class="docutils literal notranslate"><span class="pre">f&quot;{groupby_attribute}_rank&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.run_resolution_sweep_parallel">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">run_resolution_sweep_parallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolutions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0.2,</span> <span class="pre">0.4,</span> <span class="pre">0.6,</span> <span class="pre">0.8,</span> <span class="pre">1.0]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.run_resolution_sweep_parallel" title="Link to this definition"></a></dt>
<dd><p>Perform a parallelized Leiden clustering sweep across multiple resolutions.</p>
<p>This function utilizes multi-core processing to accelerate the computation of
multiple clustering resolutions. It uses the 2026 standard ‘igraph’ flavor
to ensure performance and compatibility. Each resolution is stored as a
new column in <cite>adata.obs</cite>.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>adata (AnnData): Annotated data matrix. Must have a neighbors graph</dt><dd><p>pre-computed in <cite>.obsp[‘connectivities’]</cite>.</p>
</dd>
<dt>resolutions (list of float): Resolutions to evaluate.</dt><dd><p>Default is [0.2, 0.4, 0.6, 0.8, 1.0].</p>
</dd>
</dl>
<p>n_jobs (int): Number of CPU cores to use. -1 uses all available processors.</p>
</dd>
<dt>Returns:</dt><dd><p>AnnData: The input object with added columns in <cite>.obs</cite> (e.g., ‘leiden_0.2’).</p>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k_processed</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Ensure neighbors are computed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Run parallel sweep</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">resolutions</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">run_resolution_sweep_parallel</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolutions</span><span class="o">=</span><span class="n">resolutions</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Visualize one of the results</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;leiden_0.5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.subset_and_clean_adata">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">subset_and_clean_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.subset_and_clean_adata" title="Link to this definition"></a></dt>
<dd><p>Extract a subset of cells from an AnnData object and clean categorical metadata.</p>
<p>This function filters the input AnnData by one or more keys in a specified
<code class="docutils literal notranslate"><span class="pre">.obs</span></code> column. It automatically removes unused categories from all
categorical columns in the resulting subset to prevent downstream errors
during aggregation (e.g., PyDESeq2 or pseudobulking) and plotting.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The input AnnData object to be subsetted.</p></li>
<li><p><strong>obs_col</strong> (<em>str</em>) – The column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used for filtering (e.g., ‘cluster_name’).</p></li>
<li><p><strong>keys</strong> (<em>str</em><em> | </em><em>List</em><em>[</em><em>str</em><em>]</em>) – A single string or a list of strings representing the categories to keep.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copied subset of the original AnnData containing only the specified cells.
Returns <code class="docutils literal notranslate"><span class="pre">None</span></code> if no cells match the provided keys.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Optional[AnnData]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <code class="docutils literal notranslate"><span class="pre">obs_col</span></code> is not found in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Extract a single cell type</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myo_adata</span> <span class="o">=</span> <span class="n">extract_cell_cluster_pro</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;cluster_name&quot;</span><span class="p">,</span> <span class="s2">&quot;Myo&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Combine multiple cluster IDs into one subset</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">epi_adata</span> <span class="o">=</span> <span class="n">extract_cell_cluster_pro</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;cluster-label&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;12&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.tl.summarize_category_distribution">
<span class="sig-prename descclassname"><span class="pre">scoriapy.tl.</span></span><span class="sig-name descname"><span class="pre">summarize_category_distribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">category_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.tl.summarize_category_distribution" title="Link to this definition"></a></dt>
<dd><p>Summarize the distribution of a categorical variable across groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object containing the relevant categorical data in <code class="docutils literal notranslate"><span class="pre">obs</span></code>.</p></li>
<li><p><strong>category_col</strong> (<em>str</em>) – Name of the column in <code class="docutils literal notranslate"><span class="pre">obs</span></code> whose distribution is summarized.</p></li>
<li><p><strong>group_col</strong> – Column name or list of column names in <code class="docutils literal notranslate"><span class="pre">obs</span></code> defining groups.</p></li>
<li><p><strong>selected_groups</strong> (<em>list</em><em>[</em><em>str</em><em>] </em><em>| </em><em>None</em>) – Optional list of column-prefixes to keep (e.g. <code class="docutils literal notranslate"><span class="pre">[&quot;Y&quot;,</span> <span class="pre">&quot;M&quot;,</span> <span class="pre">&quot;O&quot;]</span></code>).</p></li>
<li><p><strong>normalize</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, express counts as percentages per column and add a
<code class="docutils literal notranslate"><span class="pre">&quot;Total&quot;</span></code> row.</p></li>
<li><p><strong>sort_order</strong> (<em>list</em><em>[</em><em>str</em><em>] </em><em>| </em><em>None</em>) – Optional list defining the ordering of group prefixes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>When <code class="docutils literal notranslate"><span class="pre">normalize</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, returns a styled percentage table;
otherwise a styled count table with an extra <code class="docutils literal notranslate"><span class="pre">&quot;Total&quot;</span></code> column.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.io.formats.style.Styler or pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scoriapy.pl">
<span id="plotting-scoriapy-pl"></span><h2>Plotting (<a class="reference internal" href="generated/scoriapy.pl.html#module-scoriapy.pl" title="scoriapy.pl"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.pl</span></code></a>)<a class="headerlink" href="#module-scoriapy.pl" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.add_ncells_annotation">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">add_ncells_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.98,</span> <span class="pre">0.02)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bbox</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.add_ncells_annotation" title="Link to this definition"></a></dt>
<dd><p>Decorator that adds the total cell count (n = XXX) to a Scanpy plotting function.</p>
<p>This decorator automatically extracts the observation count from the AnnData
object and places a text annotation on the resulting axis. It forces
<cite>show=False</cite> on the wrapped function to allow for the annotation to be
added before the figure is displayed.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>loc (tuple): XY coordinates in axes-relative units (0 to 1).</dt><dd><p>Default is (0.98, 0.02) (bottom right).</p>
</dd>
</dl>
<p>fontsize (int): Size of the annotation text.
color (str): Color of the annotation text.
bbox (bool): If True, adds a semi-transparent white background box</p>
<blockquote>
<div><p>behind the text for readability.</p>
</div></blockquote>
</dd>
<dt>Returns:</dt><dd><p>function: The wrapped plotting function.</p>
</dd>
<dt>Example:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Standard usage as a decorator</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nd">@add_ncells_annotation</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="p">(</span><span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span><span class="w"> </span><span class="nf">plot_custom_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Using it directly on an existing Scanpy function</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">annotated_umap</span> <span class="o">=</span> <span class="n">add_ncells_annotation</span><span class="p">()(</span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">annotated_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cell_type&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.add_text_labels">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">add_text_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">median_positions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_props_kw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.add_text_labels" title="Link to this definition"></a></dt>
<dd><p>Add text labels at median positions on a matplotlib Axes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> – Target matplotlib Axes.</p></li>
<li><p><strong>median_positions</strong> – Mapping from label to <code class="docutils literal notranslate"><span class="pre">(x,</span> <span class="pre">y)</span></code> coordinate.</p></li>
<li><p><strong>text_props_kw</strong> (<em>dict</em><em> | </em><em>None</em>) – Optional dict of keyword arguments forwarded to <code class="docutils literal notranslate"><span class="pre">ax.text</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.calculate_median_position">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">calculate_median_position</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_subset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlay_attribute</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.calculate_median_position" title="Link to this definition"></a></dt>
<dd><p>Calculate median positions of clusters in an embedding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata_subset</strong> – AnnData object with an embedding in <code class="docutils literal notranslate"><span class="pre">.obsm[basis]</span></code>.</p></li>
<li><p><strong>basis</strong> (<em>str</em>) – Key of the embedding in <code class="docutils literal notranslate"><span class="pre">adata_subset.obsm</span></code>.</p></li>
<li><p><strong>overlay_attribute</strong> (<em>str</em>) – Observation column defining the groups whose medians are computed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapping from label to a 2D median coordinate.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.create_heatmap_annotation">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">create_heatmap_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">col_df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.create_heatmap_annotation" title="Link to this definition"></a></dt>
<dd><p>Create a <code class="xref py py-class docutils literal notranslate"><span class="pre">PyComplexHeatmap.HeatmapAnnotation</span></code> from a column DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>col_df</strong> (<em>DataFrame</em>) – Column annotation DataFrame whose columns will be turned into
annotation tracks.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Configured annotation object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>PyComplexHeatmap.HeatmapAnnotation</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.get_heatmap_dfs">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">get_heatmap_dfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.get_heatmap_dfs" title="Link to this definition"></a></dt>
<dd><p>Prepare matrices and annotation DataFrames for heatmap plotting.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with expression data.</p></li>
<li><p><strong>obs_cols</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of observation columns to include as column annotations.</p></li>
<li><p><strong>var_cols</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of variable columns to include as row annotations.</p></li>
<li><p><strong>scale</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, z-score rows of the expression matrix.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary with keys <code class="docutils literal notranslate"><span class="pre">&quot;df_mat&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;col_df&quot;</span></code>, and <code class="docutils literal notranslate"><span class="pre">&quot;row_df&quot;</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_cluster_heatmap">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_cluster_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">15)</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_cluster_heatmap" title="Link to this definition"></a></dt>
<dd><p>Plot a clustered heatmap with annotations for genes and cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_mat</strong> (<em>DataFrame</em>) – Expression matrix (rows = genes, columns = cells).</p></li>
<li><p><strong>col_df</strong> (<em>DataFrame</em>) – Column annotation DataFrame (index should align with columns of
<code class="docutils literal notranslate"><span class="pre">df_mat</span></code>).</p></li>
<li><p><strong>row_df</strong> (<em>DataFrame</em>) – Row annotation DataFrame (index should align with rows of
<code class="docutils literal notranslate"><span class="pre">df_mat</span></code> and include a <code class="docutils literal notranslate"><span class="pre">&quot;gene_cat&quot;</span></code> column).</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of the matplotlib figure.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments forwarded to
<code class="xref py py-class docutils literal notranslate"><span class="pre">PyComplexHeatmap.ClusterMapPlotter</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_distribution">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_distribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_intercept</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_distribution" title="Link to this definition"></a></dt>
<dd><p>Plot a KDE of values in <code class="docutils literal notranslate"><span class="pre">df</span></code> with an optional vertical cutoff line.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>DataFrame</em>) – DataFrame whose numeric values to visualize.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – Maximum number of columns to sample for plotting.</p></li>
<li><p><strong>x_intercept</strong> (<em>float</em>) – Location of the vertical reference line.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_embeddings_with_labels">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_embeddings_with_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlay_attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subplot_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_props_kw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_embeddings_with_labels" title="Link to this definition"></a></dt>
<dd><p>Plot embeddings with median-position labels, optionally split by a grouping column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with an embedding in <code class="docutils literal notranslate"><span class="pre">.obsm[basis]</span></code>.</p></li>
<li><p><strong>basis</strong> (<em>str</em>) – Key in <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code> for the embedding (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;X_umap&quot;</span></code>).</p></li>
<li><p><strong>color_attribute</strong> (<em>str</em>) – Observation column used to color cells.</p></li>
<li><p><strong>overlay_attribute</strong> (<em>str</em>) – Observation column used to define label groups.</p></li>
<li><p><strong>group_col</strong> (<em>str</em><em> | </em><em>None</em>) – Optional column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used to split the data into
separate subplots.</p></li>
<li><p><strong>subplot_width</strong> (<em>float</em>) – Base width of each subplot (in inches).</p></li>
<li><p><strong>text_props_kw</strong> (<em>dict</em><em> | </em><em>None</em>) – Optional dict of keyword arguments forwarded to <code class="docutils literal notranslate"><span class="pre">ax.text</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_group_embeddings">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_group_embeddings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_pca_umap'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_group_embeddings" title="Link to this definition"></a></dt>
<dd><p>Create embedding plots for each group in <code class="docutils literal notranslate"><span class="pre">group_col</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with an embedding in <code class="docutils literal notranslate"><span class="pre">.obsm[basis]</span></code>.</p></li>
<li><p><strong>group_col</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used to define groups.</p></li>
<li><p><strong>color</strong> (<em>str</em>) – Observation column used to color cells.</p></li>
<li><p><strong>basis</strong> (<em>str</em>) – Key in <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code> containing the embedding coordinates.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments forwarded to <code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.pl.embedding()</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_pca_for_samples">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_pca_for_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_pca_for_samples" title="Link to this definition"></a></dt>
<dd><p>Plot PCA embeddings for a list of sample IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object containing a <code class="docutils literal notranslate"><span class="pre">&quot;sample&quot;</span></code> column in <code class="docutils literal notranslate"><span class="pre">obs</span></code>.</p></li>
<li><p><strong>sample_list</strong> – Iterable of sample IDs to visualize.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_umap_by_categories">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_umap_by_categories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_columns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_to_cell_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'on_data'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_umap_by_categories" title="Link to this definition"></a></dt>
<dd><p>Plots embeddings (e.g., UMAP) for each unique combination of categories in specified columns of adata.obs,
ensuring consistent color mapping, optional sorting, and a single global legend.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>obs_columns</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – Columns in adata.obs to group by (e.g., [“age”, “sample”]).</p></li>
<li><p><strong>basis</strong> (<em>str</em>) – Embedding basis in adata.obsm (e.g., “umap” or “X_umap”).</p></li>
<li><p><strong>color</strong> (<em>str</em>) – Observation column used for coloring (e.g., “leiden” or “cluster_name”).</p></li>
<li><p><strong>ncol</strong> (<em>int</em>) – Number of columns in the subplot layout.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em>) – Base figure size (width, height).</p></li>
<li><p><strong>sort_order</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Custom order for sorting subsets (e.g., [“Y”, “M”, “O”]).</p></li>
<li><p><strong>cluster_to_cell_mapping</strong> (<em>dict</em><em>, </em><em>optional</em>) – Mapping from cluster names to cell type labels.</p></li>
<li><p><strong>legend_location</strong> (<em>str</em>) – <dl class="simple">
<dt>Controls legend placement:</dt><dd><ul>
<li><p>”on_data” → Cluster labels overlaid on embedding.</p></li>
<li><p>”right” → Individual legends on each subplot.</p></li>
<li><p>”none” → Only a single combined legend below all subplots.</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>point_size</strong> (<em>float</em><em>, </em><em>optional</em>) – Point size for scatter plot. Defaults to Scanpy’s size formula.</p></li>
<li><p><strong>return_fig</strong> (<em>bool</em>) – If True, returns (fig, axes) instead of displaying the figure.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_umap_by_categories_v1">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_umap_by_categories_v1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_columns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_to_cell_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'on_data'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_umap_by_categories_v1" title="Link to this definition"></a></dt>
<dd><p>Plot UMAPs for each combination of categories in <code class="docutils literal notranslate"><span class="pre">obs_columns</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object containing an embedding in <code class="docutils literal notranslate"><span class="pre">.obsm[basis]</span></code>.</p></li>
<li><p><strong>obs_columns</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em>) – Sequence of observation columns whose unique combinations define
subplots (e.g. <code class="docutils literal notranslate"><span class="pre">[&quot;age&quot;,</span> <span class="pre">&quot;sample&quot;]</span></code>).</p></li>
<li><p><strong>basis</strong> (<em>str</em>) – Key in <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code> containing the embedding coordinates.</p></li>
<li><p><strong>color</strong> (<em>str</em>) – Observation column used to color cells (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;leiden&quot;</span></code>).</p></li>
<li><p><strong>ncol</strong> (<em>int</em>) – Number of subplots per row.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Base figure size; actual height scales with the number of rows.</p></li>
<li><p><strong>sort_order</strong> (<em>Sequence</em><em>[</em><em>str</em><em>] </em><em>| </em><em>None</em>) – Optional ordering for the first component of the combined group name.</p></li>
<li><p><strong>cluster_to_cell_mapping</strong> (<em>Mapping</em><em>[</em><em>str</em><em>, </em><em>str</em><em>] </em><em>| </em><em>None</em>) – Optional mapping from cluster IDs to cell type labels for the legend.</p></li>
<li><p><strong>legend_location</strong> (<em>str</em>) – Where to place legends (<code class="docutils literal notranslate"><span class="pre">&quot;on_data&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;right&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;none&quot;</span></code>).</p></li>
<li><p><strong>point_size</strong> (<em>float</em><em> | </em><em>None</em>) – Optional marker size. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, a size proportional to the total
cell count is used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_umap_by_group">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_umap_by_group</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_column</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_umap_by_group" title="Link to this definition"></a></dt>
<dd><p>Plot UMAPs for each unique value in <code class="docutils literal notranslate"><span class="pre">group_column</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with a computed UMAP embedding.</p></li>
<li><p><strong>group_column</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> whose unique values define subplots.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments forwarded to <code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.pl.umap()</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_umap_grid_by_sample">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_umap_grid_by_sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_per_row</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize_per_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4,</span> <span class="pre">4)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_umap_grid_by_sample" title="Link to this definition"></a></dt>
<dd><p>Plot a grid of UMAPs, one per sample in <code class="docutils literal notranslate"><span class="pre">adata.obs['sample_name']</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with a UMAP embedding.</p></li>
<li><p><strong>col_per_row</strong> (<em>int</em>) – Number of UMAP subplots per row.</p></li>
<li><p><strong>figsize_per_plot</strong> (<em>tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Size of each individual UMAP subplot.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Figure containing the UMAP grid.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_umap_param_grid">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_umap_param_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.1,</span> <span class="pre">1,</span> <span class="pre">2)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spreads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.5,</span> <span class="pre">1,</span> <span class="pre">5)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_umap_param_grid" title="Link to this definition"></a></dt>
<dd><p>Plot a grid of UMAPs for all combinations of <code class="docutils literal notranslate"><span class="pre">min_dist</span></code> and <code class="docutils literal notranslate"><span class="pre">spread</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with precomputed neighbors/PCA, ready for UMAP.</p></li>
<li><p><strong>min_dists</strong> (<em>Sequence</em><em>[</em><em>float</em><em>]</em>) – Sequence of <code class="docutils literal notranslate"><span class="pre">min_dist</span></code> values to evaluate.</p></li>
<li><p><strong>spreads</strong> (<em>Sequence</em><em>[</em><em>float</em><em>]</em>) – Sequence of <code class="docutils literal notranslate"><span class="pre">spread</span></code> values to evaluate.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.plot_violin_genes">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">plot_violin_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">common_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col_adata1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col_adata2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subplot_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.plot_violin_genes" title="Link to this definition"></a></dt>
<dd><p>Plot violin plots for genes across specified groups from two AnnData objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata1</strong> – First AnnData object.</p></li>
<li><p><strong>adata2</strong> – Second AnnData object.</p></li>
<li><p><strong>common_genes</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of genes to visualize.</p></li>
<li><p><strong>group_col_adata1</strong> (<em>str</em>) – Grouping column in <code class="docutils literal notranslate"><span class="pre">adata1.obs</span></code>.</p></li>
<li><p><strong>group_col_adata2</strong> (<em>str</em>) – Grouping column in <code class="docutils literal notranslate"><span class="pre">adata2.obs</span></code>.</p></li>
<li><p><strong>subplot_width</strong> (<em>float</em>) – Width of each subplot in inches.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.preserve_leiden_categories">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">preserve_leiden_categories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_sub</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_full</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leiden_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'leiden'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.preserve_leiden_categories" title="Link to this definition"></a></dt>
<dd><p>Ensures that the leiden column in a subset AnnData object retains the full category order
and color mapping from the original AnnData object.</p>
<p>Parameters:
- adata_sub: Subset AnnData object.
- adata_full: Full AnnData object (before subsetting).
- leiden_column: The name of the leiden clustering column in <cite>adata.obs</cite>.</p>
<p>Returns:
- Modified <cite>adata_sub</cite> with preserved leiden categories and color mapping.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.sort_cells_by_metadata">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">sort_cells_by_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_sort_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primary_categories</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_sort_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_categories</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.sort_cells_by_metadata" title="Link to this definition"></a></dt>
<dd><p>Sort cells by two categorical obs columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object to sort.</p></li>
<li><p><strong>primary_sort_col</strong> (<em>str</em>) – Name of the primary categorical column (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;age&quot;</span></code>).</p></li>
<li><p><strong>primary_categories</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Desired ordering of categories in <code class="docutils literal notranslate"><span class="pre">primary_sort_col</span></code>.</p></li>
<li><p><strong>second_sort_col</strong> (<em>str</em>) – Name of the secondary categorical column.</p></li>
<li><p><strong>second_categories</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Desired ordering of categories in <code class="docutils literal notranslate"><span class="pre">second_sort_col</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>AnnData object reordered according to the sorted index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.pl.subset_adata">
<span class="sig-prename descclassname"><span class="pre">scoriapy.pl.</span></span><span class="sig-name descname"><span class="pre">subset_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes_map</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.pl.subset_adata" title="Link to this definition"></a></dt>
<dd><p>Subset AnnData by selected groups and genes, assigning gene categories.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object with a populated <code class="docutils literal notranslate"><span class="pre">raw</span></code> attribute.</p></li>
<li><p><strong>group_col</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used to select cells.</p></li>
<li><p><strong>keys</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – Values in <code class="docutils literal notranslate"><span class="pre">group_col</span></code> that define the subset of cells.</p></li>
<li><p><strong>genes_map</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – Mapping from gene name to category label. Only genes present in
<code class="docutils literal notranslate"><span class="pre">adata.raw.var_names</span></code> are used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Subsetted AnnData with selected genes and cells; gene categories are
stored in <code class="docutils literal notranslate"><span class="pre">.var['gene_cat']</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-scoriapy.utils">
<span id="utilities-scoriapy-utils"></span><h2>Utilities (<a class="reference internal" href="generated/scoriapy.utils.html#module-scoriapy.utils" title="scoriapy.utils"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scoriapy.utils</span></code></a>)<a class="headerlink" href="#module-scoriapy.utils" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="scoriapy.utils.FigureRegistry">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">FigureRegistry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'figures'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.FigureRegistry" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A lightweight container for managing, annotating, displaying, and batch-saving matplotlib figures in Jupyter workflows.</p>
<p>Features:
- Tracks figures by name (and optional tags) with creation timestamps.
- Batch-saving of all figures (e.g., as PDFs) to a chosen output folder.
- Convenient <cite>.show()</cite> gallery for Jupyter (with names/tags above each figure).
- Automatic rasterization of collections to control output file size.
- Easy repr for auditing what’s stored.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory to save figures (default “figures”).</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – DPI for figure saving (default 300).</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">add(name,</span> <span class="pre">fig=None,</span> <span class="pre">rasterize=True,</span> <span class="pre">tags=None):</span></span></dt>
<dd><p>Add a figure (current or supplied) to the registry, with optional tags and rasterization.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">save_all(overwrite=True):</span></span></dt>
<dd><p>Save all registered figures to disk (default: PDF).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">show():</span></span></dt>
<dd><p>Display all figures in a Jupyter notebook cell
(with labels and tags).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">list_figures():</span></span></dt>
<dd><p>List all tracked figure names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">__repr__():</span></span></dt>
<dd><p>Pretty print a summary table of tracked figures.</p>
</dd></dl>

<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reg</span> <span class="o">=</span> <span class="n">FigureRegistry</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;myfigs&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Add a figure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">9</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reg</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;parabola&quot;</span><span class="p">,</span> <span class="n">fig</span><span class="p">,</span> <span class="n">tags</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;demo&quot;</span><span class="p">,</span> <span class="s2">&quot;math&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Show all figures in Jupyter</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reg</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Save all to PDF in &quot;myfigs&quot; directory</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reg</span><span class="o">.</span><span class="n">save_all</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">reg</span><span class="p">)</span>
<span class="go">Name                      | Tags                 | Created</span>
<span class="go">----------------------------------------------------------------------</span>
<span class="go">parabola                  | demo, math           | 2026-01-07 16:24:33</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="scoriapy.utils.FigureRegistry.add">
<span class="sig-name descname"><span class="pre">add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rasterize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.FigureRegistry.add" title="Link to this definition"></a></dt>
<dd><p>Add figure and metadata to container.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scoriapy.utils.FigureRegistry.save_all">
<span class="sig-name descname"><span class="pre">save_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.FigureRegistry.save_all" title="Link to this definition"></a></dt>
<dd><p>Saves all figures as PDFs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="scoriapy.utils.FigureRegistry.show">
<span class="sig-name descname"><span class="pre">show</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.FigureRegistry.show" title="Link to this definition"></a></dt>
<dd><p>Visual gallery for Jupyter: displays every figure with its name and tags.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.confirm_and_set_colors">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">confirm_and_set_colors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_map</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.confirm_and_set_colors" title="Link to this definition"></a></dt>
<dd><p>Interactively confirm and apply a color map for a categorical obs column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object.</p></li>
<li><p><strong>color_attribute</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used for coloring.</p></li>
<li><p><strong>color_map</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – Mapping from category to color code.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.create_cell_mask">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">create_cell_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.create_cell_mask" title="Link to this definition"></a></dt>
<dd><p>Create a boolean mask for cells based on multiple flexible conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The AnnData object.</p></li>
<li><p><strong>conditions</strong> (<em>dict</em>) – <p>A mapping of <code class="docutils literal notranslate"><span class="pre">obs</span></code> column name to a condition. The condition can be:</p>
<ul>
<li><p>a single value (<code class="docutils literal notranslate"><span class="pre">str</span></code>, <code class="docutils literal notranslate"><span class="pre">int</span></code>, <code class="docutils literal notranslate"><span class="pre">float</span></code>)</p></li>
<li><p>a list/tuple/array/Series of allowed values</p></li>
<li><p>a callable that takes a Series and returns a boolean Series</p></li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean mask over observations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.display_colors">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">display_colors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">color_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.display_colors" title="Link to this definition"></a></dt>
<dd><p>Display colors as inline blocks in a notebook.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>color_list</strong> – Iterable of color specifications (e.g. hex strings) to display.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.display_colors_with_keys">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">display_colors_with_keys</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">color_map</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.display_colors_with_keys" title="Link to this definition"></a></dt>
<dd><p>Display a mapping of keys to colors as inline blocks in a notebook.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>color_map</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – Mapping from label to color code to visualize.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.extract_cell_cluster">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">extract_cell_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_obs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.extract_cell_cluster" title="Link to this definition"></a></dt>
<dd><p>Extract a subset of cells from an AnnData object based on values in <code class="docutils literal notranslate"><span class="pre">.obs</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object.</p></li>
<li><p><strong>col_obs</strong> (<em>str</em>) – Column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> used for filtering.</p></li>
<li><p><strong>keys</strong> – Single key or list of keys to keep.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Subsetted AnnData or <code class="docutils literal notranslate"><span class="pre">None</span></code> if no cells match.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.get_full_df">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">get_full_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.get_full_df" title="Link to this definition"></a></dt>
<dd><p>Extract expression data from <code class="docutils literal notranslate"><span class="pre">adata.raw</span></code> and optionally z-score scale it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object.</p></li>
<li><p><strong>keys</strong> (<em>list</em><em>[</em><em>str</em><em>] </em><em>| </em><em>None</em>) – Variable names to extract. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all variables are used.</p></li>
<li><p><strong>scale</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, apply z-score column-wise.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Expression matrix (optionally scaled).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.get_mat_df">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">get_mat_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.get_mat_df" title="Link to this definition"></a></dt>
<dd><p>Extract expression data as a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object.</p></li>
<li><p><strong>keys</strong> (<em>list</em><em>[</em><em>str</em><em>] </em><em>| </em><em>None</em>) – Variable names to extract. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all variables are used.</p></li>
<li><p><strong>use_raw</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, use <code class="docutils literal notranslate"><span class="pre">adata.raw</span></code>; otherwise use <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Expression matrix with observations as rows.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.make_annotation_df">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">make_annotation_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_dict</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.make_annotation_df" title="Link to this definition"></a></dt>
<dd><p>Combine metadata from multiple AnnData objects into a single DataFrame.</p>
<p>Currently expects keys <code class="docutils literal notranslate"><span class="pre">&quot;age&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;rapa&quot;</span></code> and normalizes the
column name to <code class="docutils literal notranslate"><span class="pre">&quot;group&quot;</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>adata_dict</strong> (<em>dict</em>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.make_mat_dfs">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">make_mat_dfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">common_genes</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.make_mat_dfs" title="Link to this definition"></a></dt>
<dd><p>Extract scaled expression matrices for a list of common genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata_dict</strong> (<em>dict</em>) – Mapping of dataset name to AnnData.</p></li>
<li><p><strong>common_genes</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – Genes to extract.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapping from dataset name to a transposed expression DataFrame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.map_colors">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">map_colors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapping_dict</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.map_colors" title="Link to this definition"></a></dt>
<dd><p>Map colors from a reference dictionary to a target dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – Mapping from reference category to color (e.g. cluster ID → hex code).</p></li>
<li><p><strong>target_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – Mapping whose keys define the target categories to color.
Existing values are ignored; only the keys are used.</p></li>
<li><p><strong>mapping_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>Iterable</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Mapping from reference keys to iterable(s) of target keys. Each
target key inherits the color of its reference key.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>New mapping from target keys to colors. Keys not specified in
<code class="docutils literal notranslate"><span class="pre">mapping_dict</span></code> are assigned gray (<code class="docutils literal notranslate"><span class="pre">&quot;#808080&quot;</span></code>).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.scale_df">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">scale_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.scale_df" title="Link to this definition"></a></dt>
<dd><p>Apply z-score normalization to each column of a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>DataFrame</em>) – Input DataFrame.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with standardized columns (mean=0, std=1).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.setup_logger">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">setup_logger</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">log_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.setup_logger" title="Link to this definition"></a></dt>
<dd><p>Create and configure a logger for Scanpy processing pipelines.</p>
<p>This logger:
- prints messages to the console with timestamps
- optionally writes logs to a file
- avoids duplicate handlers (important in Jupyter notebooks)
- supports adjustable log levels (INFO, DEBUG, WARNING, etc.)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>log_file</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – If provided, log messages are also written to this file.
If None, only console output is produced.</p></li>
<li><p><strong>level</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: logging.INFO</em><em>)</em>) – Logging verbosity level. Typically one of:
- logging.DEBUG
- logging.INFO
- logging.WARNING
- logging.ERROR
- logging.CRITICAL</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>logger</strong> – A configured logger instance that can be passed to functions
(e.g., preprocessing pipelines) to produce timestamped logs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>logging.Logger</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Inside Jupyter notebooks, calling this function repeatedly will not
produce duplicate log messages because the logger only adds handlers
the first time it is constructed.</p>
<p class="rubric">Examples</p>
<p>Basic usage:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span> <span class="o">=</span> <span class="n">setup_logger</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting preprocessing...&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Logging to both console and file:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span> <span class="o">=</span> <span class="n">setup_logger</span><span class="p">(</span><span class="s2">&quot;scanpy_run.log&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running HVG selection...&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Using inside a Scanpy pipeline:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span> <span class="o">=</span> <span class="n">setup_logger</span><span class="p">(</span><span class="s2">&quot;fibro_preprocess.log&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">run_scanpy_basic_pipeline</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Pipeline finished.&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Changing verbosity:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span> <span class="o">=</span> <span class="n">setup_logger</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">DEBUG</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Debug-level message for troubleshooting.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.subset_adata_by_mask">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">subset_adata_by_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_mask</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.subset_adata_by_mask" title="Link to this definition"></a></dt>
<dd><p>Subset an AnnData object using a cell mask and gene list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The original AnnData object.</p></li>
<li><p><strong>cell_mask</strong> (<em>ndarray</em>) – Boolean mask for selecting cells.</p></li>
<li><p><strong>gene_list</strong> (<em>list</em>) – Genes to keep.</p></li>
<li><p><strong>use_raw</strong> (<em>bool</em>) – Whether to use <code class="docutils literal notranslate"><span class="pre">adata.raw</span></code> for gene subsetting.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new subsetted AnnData object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scoriapy.utils.subset_by_markers">
<span class="sig-prename descclassname"><span class="pre">scoriapy.utils.</span></span><span class="sig-name descname"><span class="pre">subset_by_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strategy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'any'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">require_n_markers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_obs_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#scoriapy.utils.subset_by_markers" title="Link to this definition"></a></dt>
<dd><p>Subset cells based on expression of a set of marker genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object containing the expression matrix.</p></li>
<li><p><strong>markers</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – List of marker gene names. Genes not found in <code class="docutils literal notranslate"><span class="pre">adata.var_names</span></code> are
silently ignored.</p></li>
<li><p><strong>strategy</strong> (<em>Literal</em><em>[</em><em>'any'</em><em>, </em><em>'all'</em><em>, </em><em>'count'</em><em>, </em><em>'threshold'</em><em>]</em>) – <p>Strategy used to define positive cells:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;any&quot;</span></code> – keep cells expressing <em>any</em> marker above <code class="docutils literal notranslate"><span class="pre">threshold</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code> – keep cells expressing <em>all</em> markers above <code class="docutils literal notranslate"><span class="pre">threshold</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;count&quot;</span></code> – keep cells with at least <code class="docutils literal notranslate"><span class="pre">min_count</span></code> markers above
<code class="docutils literal notranslate"><span class="pre">threshold</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;threshold&quot;</span></code> – keep cells whose <strong>mean</strong> marker expression
exceeds <code class="docutils literal notranslate"><span class="pre">threshold</span></code>.</p></li>
</ul>
</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – Expression cutoff to consider a marker as expressed.</p></li>
<li><p><strong>require_n_markers</strong> (<em>int</em>) – Minimum number of markers required to be present in
<code class="docutils literal notranslate"><span class="pre">adata.var_names</span></code>. Currently this is advisory only and does not alter
behaviour, but can be used by callers for validation.</p></li>
<li><p><strong>min_count</strong> (<em>int</em>) – Minimum number of expressed markers when <code class="docutils literal notranslate"><span class="pre">strategy='count'</span></code>.</p></li>
<li><p><strong>use_raw</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> and <code class="docutils literal notranslate"><span class="pre">adata.raw</span></code> is present, use <code class="docutils literal notranslate"><span class="pre">adata.raw.X</span></code> instead
of <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>.</p></li>
<li><p><strong>add_obs_key</strong> (<em>str</em><em> | </em><em>None</em>) – Optional name of an <code class="docutils literal notranslate"><span class="pre">obs</span></code> column to store the boolean mask.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Copy of <code class="docutils literal notranslate"><span class="pre">adata</span></code> restricted to cells that satisfy the marker rule.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="ScoriaPy documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2026, ScoriaPy contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>